### PR TITLE
MBR and GPT creation support

### PR DESCRIPTION
**Pull request by dxdxdt**
~**Please DO NOT merge** this draft PR yet.~

This is long overdue, but it's a massive change. I intend to do RFC in the kernel list to first to see if the heuristics I made using sysfs is the right approach. Any RFC from Linux gurus welcome.

dosfstools was originally written in the 90's and a lot in Linux has changed since than. I think that we no longer have to rely on symlink dereferencing and sysfs tree walking to figure out if the device is "real", "virtual" or whatnot. So many things could go wrong doing those things in C.

So I propose using just a few files and symlinks in `/sys/dev/block/*/`. See changes in `exfat_get_blk_dev_info()` and `exfat_select_part_type()`. More on that in the list thread later on.

`mkfs.fat` and `mkfs.exfat` have to support creation of partition table all because of MS Windows. Consider a scenario where the user formats a medium with `mkfs.exfat`, dumps some data to it, and then some time later, they realise that the medium cannot be used on Windows. This is a serious problem. This PR addresses it.

To make such medium usable on Windows, if the capacity is less than 2TB, a recursive MBR partion can be placed in the exFAT bootsector region so that Windows can read it and scan the volume. This is fairly easy to implement and of low-risk because only two blocks(main and back up) has to be backed up before manipulating the bootsectors.

However, for devices larger than 2TB, there's no clean cut way to do it because GPT structures take up more than just the first 512 bytes(LBA0). The exFAT headers and cluster data need to be migrated to place GPT structures. It's just now worth it and that's why I went on to teach mkfs.exfat how to make GPT so that users would never have to be affected by the problem in the first place.

By supporting GPT in-house, users using `mkfs.exfat` directly or indirectly(say on Android, or other GUI tools) won't have to care about all the MBR/GPT/removable/fixed shenanigans. Just use the tool and everything will be set up properly.

---

https://github.com/dxdxdt/exfatprogs/actions/runs/25025039365

Windows CHKDSK util reports wrong volume size. Just a userspace bug. Nothing serious.

Screenshot_win2k25_2026-04-28_05:41:15

Screenshot_20260426_114859

MBR bootstrap code taken and improved from dosfstools.

Fixes: https://github.com/exfatprogs/exfatprogs/issues/348
Fixes: https://github.com/exfatprogs/exfatprogs/issues/261
Fixes: https://github.com/exfatprogs/exfatprogs/issues/267
Merge after: #350, on 12/05/2026 or onwards(2 weeks review period ;))

```
David Timber (16):
  mkfs: print volume UUID
  mkfs: default cluster size to 512 for small devices
  Fix exfat_read_sector() and exfat_write_sector() for 64-bit LBA
  Do not run on pre-2.6 Linux kernel
  Fix print_guid()
  Include release date in version line
  mkfs: use dev_fd to do O_DIRECT read
  manpages: fix up mkfs.exfat man page format regarding --upcase option
  mkfs: add partition table creation support for Windows compatibility
  Github workflow: test mkfs.exfat -P and -C option
  Github workflow: add Valgrind tests
  mkfs: catch corner cases of parse_size()
  Refactor MBR manipulation and partition type selection heuristics
  fsck: add --put-mbr and --clear-mbr options
  mkfs: adjust total cluster limit
  Refactor libexfat and minor fix in exfat2img and mkfs

 .github/workflows/c-cpp.yml | 224 +++++++++---
 defrag/defrag.c             |   9 +-
 dump/dump.c                 |  13 +-
 exfat2img/exfat2img.c       |  38 ++-
 fsck/fsck.c                 | 248 ++++++++++++--
 fsck/fsck.h                 |   1 +
 fsck/repair.c               |   2 +
 fsck/repair.h               |   2 +
 include/exfat_ondisk.h      |  18 +-
 include/libexfat.h          | 102 +++++-
 include/version.h           |   1 +
 label/label.c               |  14 +-
 lib/exfat_fs.c              |  30 +-
 lib/libexfat.c              | 609 +++++++++++++++++++++++++++++----
 manpages/fsck.exfat.8       |  27 +-
 manpages/mkfs.exfat.8       |  37 +-
 mkfs/Makefile.am            |   2 +-
 mkfs/crc.c                  |  73 ++++
 mkfs/mkfs.c                 | 663 +++++++++++++++++++++++++++++++++---
 mkfs/mkfs.h                 | 111 +++++-
 mkfs/upcase.c               |   7 +-
 tune/tune.c                 |  20 +-
 22 files changed, 1994 insertions(+), 257 deletions(-)
 create mode 100644 mkfs/crc.c

-- 
2.53.0.1.ga224b40d3f.dirty
```

### CODE REVIEW

**Comment by namjaejeon** [comment]
@hclee Please check this PR.

**Comment by dxdxdt** [comment]
https://lore.kernel.org/linux-fsdevel/d7edfeab-af06-4b2d-a085-f8eef077e877@dev.snart.me/

Nobody seems to care. I've done my part to the best of my abilities. If anyone's having issues related to the removable SCSI device detection heuristics in the future, you may reference the mailing list thread and this PR and go from there.

Signing it off for review.

**Comment by namjaejeon** [comment]
@dxdxdt there is a conflict with this patch(https://github.com/exfatprogs/exfatprogs/commit/61e1eb311d6a907b760e56258f7c276e79023e3f).

> Please DO NOT merge this draft PR yet.

Is it not ready yet ? or can merge this PR now ?

**Comment by dxdxdt** [comment]
> @dxdxdt there is a conflict with this patch([61e1eb3](https://github.com/exfatprogs/exfatprogs/commit/61e1eb311d6a907b760e56258f7c276e79023e3f)).
> 
> > Please DO NOT merge this draft PR yet.
> 
> Is it not ready yet ? or can merge this PR now ?

Oops. Sorry. Forgot to edit that out.

Sure, go for it.

**Comment by namjaejeon** [comment]
Applied it to #exfat-next.
Thanks!

File: include/version.h [hunk 1]
```diff
@@ -6,5 +6,6 @@
 #ifndef _VERSION_H
 
 #define EXFAT_PROGS_VERSION "1.3.2"
+#define EXFAT_PROGS_RELEASE_DATE "YYYY-MM-DD"
```
Comment (namjaejeon) [inline include/version.h:9]: @hclee You should update it with version number when releasing exfatprogs.
